File: lib/languages/elixir/guards.ex
# Guards in function clauses
def describe(0), do: "Zero"
def describe(n) when n > 0 and n < 10, do: "Small positive"
def describe(n) when n >= 10 and n < 100, do: "Medium positive"
def describe(n) when n >= 100, do: "Large positive"
def describe(n) when n < 0 and n > -10, do: "Small negative"
def describe(n) when n <= -10 and n > -100, do: "Medium negative"
def describe(_), do: "Large negative"

# Guards with destructuring
def process_pair({a, b}) when a == b, do: "Equal"
def process_pair({a, b}) when a > b, do: "First is larger"
def process_pair({a, b}) when a < b, do: "Second is larger"

# Guards in case expressions
def describe_list(items) do
  case items do
    [] -> "Empty list"
    [x] when x > 0 -> "Single positive item"
    [x] when x < 0 -> "Single negative item"
    [x] -> "Single zero item"
    [x, y | _] when x > y -> "Decreasing list"
    [x, y | _] when x < y -> "Increasing list"
    _ -> "Other list"
  end
end

# Guards with type checks
def process(value) when is_binary(value), do: "String: #{value}"
def process(value) when is_integer(value), do: "Integer: #{value}"
def process(value) when is_float(value), do: "Float: #{value}"
def process(value) when is_list(value), do: "List with #{length(value)} items"
def process(value) when is_map(value), do: "Map with #{map_size(value)} keys"
def process(_), do: "Other type"

# Guard limitations - only certain expressions are allowed
# Custom functions cannot be used in guards
# Only built-in functions marked as guard-safe can be used

File: lib/languages/rust/guards.rs
// Guards in match expressions
fn describe(x: i32) -> &'static str {
   match x {
       0 => "Zero",
       n if n > 0 && n < 10 => "Small positive",
       n if n >= 10 && n < 100 => "Medium positive",
       n if n >= 100 => "Large positive",
       n if n < 0 && n > -10 => "Small negative",
       n if n <= -10 && n > -100 => "Medium negative",
       _ => "Large negative",
   }
}

// Guards with destructuring
fn process_pair(pair: (i32, i32)) -> &'static str {
   match pair {
       (a, b) if a == b => "Equal",
       (a, b) if a > b => "First is larger",
       (a, b) if a < b => "Second is larger",
       _ => "Impossible case",  // (for completeness)
   }
}

// Guards in if-let expressions
fn check_positive(value: Option<i32>) -> &'static str {
   if let Some(x) = value {
       if x > 0 {
           "Positive"
       } else {
           "Zero or negative"
       }
   } else {
       "No value"
   }
}

// Guards with complex patterns
enum Message {
   Quit,
   Move { x: i32, y: i32 },
   Write(String),
   ChangeColor(i32, i32, i32),
}

fn process_message(msg: Message) -> &'static str {
   match msg {
       Message::Move { x, y } if x == 0 && y == 0 => "Stay in place",
       Message::Move { x, y } => "Move",
       Message::Write(text) if text.is_empty() => "Write nothing",
       Message::Write(_) => "Write something",
       Message::ChangeColor(r, g, b) if r == g && g == b => "Change to grayscale",
       Message::ChangeColor(_, _, _) => "Change to color",
       Message::Quit => "Quit",
   }
}

